### ISSUE TITLE
Is inline colouring possible?

### ISSUE THREAD

**Issue by xrheihwh** [AndreRenaud/PDFGen#154]
Lets say I wanted to emphasis or colourize mid part of the sentence.

[default black]This is an example[now red]of inline colouring[now black]within the same sentence in a PDF document.

I'm asking since you're supposed to position manually just like how it is with postscript (and in turn, PDF) creation, I feel it would be a pain to manually place multiple strings on the same line.

**Comment by AndreRenaud** [comment]
Hey. Yes, that would be a good feature. The absolute positioning makes that pretty tricky at the moment, so it would be better to be done automatically under the hood. It would require some kind of markdown-like parser, which might be a bit involved. Is there a simple syntax that already exists that might suit (to cover things like underline, strike through, colour etc...)?

Right now there isn't a trivial answer - you'd have to break it up yourself externally and then build each element separately with absolute positioning. Possible, but awkward.

**Comment by yslflow** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by SamWindell** [comment]
I can't think of any markup-like format that would perfect suite this library's needs exactly. A custom format might be a reasonable choice though. Something as simple as pairs of square brackets. Content in the first square brackets dictate what the style is, and content in the second is the text. This would make writing a parser easy. So: 

```
[bold][This text is bold]
[italic][This text is italic]
[size 24][This text is large]
[blue][This text is blue]
[#ff0000][This text is red]
[uri https://example.com][This text is a link]
```

**Comment by AndreRenaud** [comment]
That markdown system seems ok. It'd be nice if there was something semi-standard that we could use, rather than inventing another new one. But traditional markdown is too complex for this scenario I think.

I don't have time at the moment to implement this parser though, so I'm not sure when this will get actioned.

**Comment by AnonymouX47** [comment]
https://www.bbcode.org/ looks like a good fit, and I believe it's pretty much standardised. A subset could be implemented, and that subset specified.

**Comment by AndreRenaud** [comment]
Yes, using bbcode would make the most sense, it's very similar to what @SamWindell was suggesting already, and writing a minimal parser for it probably isn't too terrible. I might have a play with this, but very likely that I'll run out of time on it, so any PR/assistance would be welcome.

**Comment by ericoporto** [comment]
Just to mention that bbcode is indeed a good fit, I made a very simple [AGS Script module](https://github.com/ericoporto/fancy) that had a dumbed down syntax inspired by bbcode, [and the parse was very simple](https://github.com/ericoporto/fancy/blob/2252f090bdd6d29550d4c17d8ff165638d83ba84/fancy_demo/fancy.asc#L273) to implement.

**Comment by AndreRenaud** [comment]
This has been sitting here for ages, but I finally got around to sorting something out. Can now do:
```
 pdf_add_bbcode(pdf, NULL,
                   "bbcode: [b]bold[/b], [i]italic[/i], [b][i]nested "
                   "[color=#ff0000]red[/color][/i][/b], "
                   "[url=https://github.com/AndreRenaud/PDFGen]a link[/url] "
                   "and [not a tag]",
                   12, 50, 470, PDF_RGB(0, 0, 0));
```
To add bbcode encoded text.
bold/italic are supported, but not underline (these depend on the font doing it, so don't work with TTF yet.

**Comment by SamWindell** [comment]
> This has been sitting here for ages, but I finally got around to sorting something out. Can now do:
> ```
>  pdf_add_bbcode(pdf, NULL,
>                    "bbcode: [b]bold[/b], [i]italic[/i], [b][i]nested "
>                    "[color=#ff0000]red[/color][/i][/b], "
>                    "[url=https://github.com/AndreRenaud/PDFGen]a link[/url] "
>                    "and [not a tag]",
>                    12, 50, 470, PDF_RGB(0, 0, 0));
> ```
> To add bbcode encoded text.
> bold/italic are supported, but not underline (these depend on the font doing it, so don't work with TTF yet.

Incredible! Thank you.
